src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,131 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] EMPTY_DATA = {};
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        this.elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
 
     @Override
     public void add(T value) {
-
+        if (elements.length == size) {
+            grow(size + 1);
+        }
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        if (elements.length == size) {
+            grow(size + 1);
+        }
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] listElements = (T[]) new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            listElements[i] = list.get(i);
+        }
+        System.arraycopy(listElements, 0, elements, size, listElements.length);
+        size += listElements.length;
+        ;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheckForGet(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size) {
+            throw new ArrayIndexOutOfBoundsException("Insert impossible");
+        }
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheckForAdd(index);
+        T oldValue = elements[index];
+        T[] newElements = elements;
+        fastRemove(newElements, index);
+        return oldValue;
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (t == null) {
+                size--;
+                return null;
+            }
+            if (t.equals(elements[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Item not found");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        if (size == 0) {
+            return true;
+        }
         return false;
     }
+
+    private T[] grow(int minCapacity) {
+        int oldCapacity = elements.length;
+        if (oldCapacity > 0 || elements != EMPTY_DATA) {
+            int newCapacity = oldCapacity + (oldCapacity >> 1);
+            T[] newElements = (T[]) new Object[size];
+            System.arraycopy(elements, 0, newElements, 0, size);
```

Comment:
this would be enough and then you can just reassign elements to this array. There is no point in second usage of arraycopy

Reply:
done